tests/docker_mock.py

```diff
@@ -0,0 +1,77 @@
+  'Id': '57a8d9967f41ad2fdf952d1865eb4686d38931c24c70f52b7e182d795b02511a',
+  'Status': 'Exited (2) 16 hours ago'},
+ {'Created': 1430293290,
+  'Image': 'busybox:latest',
+  'Names': ['/boring_mestorf'],
+  'Id': '105026325ff668ccf4dc2bcf4f009ea35f2c6a933a778993e6fad3c50173aaab',
+  'Command': COMMAND}]
+
+mock_image = \
+    {'Created': 1414577076,
+     'Id': '3ab9a7ed8a169ab89b09fb3e12a14a390d3c662703b65b4541c0c7bde0ee97eb',
+     'ParentId': 'a79ad4dac406fcf85b9c7315fe08de5b620c1f7a12f45c8185c843f4b4a49c4e',
+     'RepoTags': ['buildroot-fedora:latest'],
+     'Size': 0,
+     'VirtualSize': 856564160}
+
+mock_logs = 'uid=0(root) gid=0(root) groups=10(wheel)'
+
+mock_pull_logs = \
+    '{"stream":"Trying to pull repository localhost:5000/busybox ..."}\r\n'\
+    '{}\r\n'\
+    '{"status":"Pulling image (latest) from localhost:5000/busybox","progressDetail":{},"id":"8c2e06607696"}'\
+    '{"status":"Pulling dependent layers","progressDetail":{},"id":"8c2e06607696"}'\
+    '{"status":"Download complete","progressDetail":{},"id":"8c2e06607696"}'\
+    '{"status":"Status: Image is up to date for localhost:5000/busybox:latest"}\r\n'
+
+mock_push_logs = \
+    '{"status":"The push refers to a repository [localhost:5000/dock-tests-b3a11e13d27c428f8fa2914c8c6a6d96] (len: 1)"}\r\n' \
+    '{"errorDetail":{"message":"Repository does not exist: localhost:5000/dock-tests-b3a11e13d27c428f8fa2914c8c6a6d96"},' \
+    '"error":"Repository does not exist: localhost:5000/dock-tests-b3a11e13d27c428f8fa2914c8c6a6d96"}\r\n'
+
+
+def mock_docker(build_should_fail=False, inspect_should_fail=False, provided_image_repotags=None):
+    if provided_image_repotags:
+        mock_image['RepoTags'] = provided_image_repotags
+    build_result = iter(['{"error":"fail"}']) if build_should_fail else iter([])
+    inspect_result = None if inspect_should_fail else mock_image
+
+    flexmock(docker.Client, build=lambda **kwargs: build_result)
+    flexmock(docker.Client, commit=lambda cid, **kwargs: mock_containers[0])
+    flexmock(docker.Client, containers=lambda **kwargs: mock_containers)
+    flexmock(docker.Client, create_container=lambda img, **kwargs: mock_containers[0])
+    flexmock(docker.Client, images=lambda **kwargs: [mock_image])
+    flexmock(docker.Client, inspect_image=lambda im_id: inspect_result)
+    flexmock(docker.Client, logs=lambda cid, **kwargs: iter([mock_logs]) if kwargs.get('stream') else mock_logs)
+    flexmock(docker.Client, pull=lambda img, **kwargs: iter([mock_pull_logs]))
+    flexmock(docker.Client, push=lambda iid, **kwargs: mock_push_logs)
+    flexmock(docker.Client, remove_container=lambda cid, **kwargs: None)
+    flexmock(docker.Client, remove_image=lambda iid, **kwargs: None)
+    flexmock(docker.Client, start=lambda cid, **kwargs: None)
+    flexmock(docker.Client, tag=lambda img, rep, **kwargs: True)
+    flexmock(docker.Client, wait=lambda cid: 1)
+    flexmock(os.path, exists=lambda path: True)  # DOCKER_SOCKET_PATH check
```

Comment:
I don't know how I can restrict the os.path.exists() mock, because something like the following leads to recursion.

flexmock(os.path, exists=lambda path: True if path == DOCKER_SOCKET_PATH else os.path.exists(path))

But so far it's OK as we use os.path.exists() only on one place where we check for DOCKER_SOCKET_PATH.